indigo_app/views/tasks.py

```diff
@@ -245,13 +245,65 @@ def post(self, request, *args, **kwargs):
                 if not has_transition_perm(state_change, self):
                     raise PermissionDenied
                 state_change(user)
-                action.send(user, verb=verb, action_object=task,
-                            place_code=task.place.place_code)
                 if verb == 'submitted':
+                    task.last_assigned_to = task.assigned_to
+                    task.assigned_to = None
+                    action.send(user, verb=verb, action_object=task,
+                                place_code=task.place.place_code)
                     messages.success(request, u"Task '%s' has been submitted for review" % task.title)
+                elif verb == 'unsubmitted':
+                    assignee = task.last_assigned_to
+                    task.assigned_to = assignee
+                    if user.id == assignee.id:
+                        action.send(user, verb='unsubmitted and picked up', action_object=task,
+                                    place_code=task.place.place_code)
+                        messages.success(request, u"You have reopened and picked up the task '%s'" % task.title)
+                    else:
+                        action.send(user, verb='unsubmitted and reassigned', action_object=task,
+                                    target=assignee,
+                                    place_code=task.place.place_code)
+                        messages.success(request, u"Task '%s' has been reopened and reassigned" % task.title)
+
                 else:
+                    action.send(user, verb=verb, action_object=task,
+                                place_code=task.place.place_code)
                     messages.success(request, u"Task '%s' has been %s" % (task.title, verb))
 
         task.save()
 
         return redirect('task_detail', place=self.kwargs['place'], pk=self.kwargs['pk'])
+
+
+class TaskAssignView(TaskViewBase, View, SingleObjectMixin):
+    # permissions
+    permission_required = ('indigo_api.change_task',)
+
+    unassign = False
+    http_method_names = [u'post']
+    model = Task
+
+    def post(self, request, *args, **kwargs):
+        task = self.get_object()
+        user = self.request.user
+        task.updated_by_user = user
+        if self.unassign:
+            task.assigned_to = None
+            action.send(user, verb='unassigned', action_object=task,
+                        place_code=task.place.place_code)
+            messages.success(request, u"Task '%s' has been unassigned" % task.title)
+        else:
+            assignee = User.objects.get(id=self.request.POST.get('user_id'))
```

Comment:
You need to double check that the user is allowed to be assigned this task. Either via a form element's queryset (so the form will validate), or manually here.